test/unit_tests/container/node.cpp

```diff
@@ -538,6 +537,123 @@ TEST_CASE("SOA-backed Node structure", "[Neuron][data_structures][node]") {
+                set_fast_imem(true);
+                check_default(node);
+                // Note there is a subtlety to the "invalid handles never become valid" rule,
+                // unfortunately. In this case, a handle was created invalid (because it refers to a
+                // valid row in a disabled column), but it became valid when that column was
+                // enabled. The rule still holds in the sense that once a handle becomes invalid
+                // because its *row* has been invalidated, it will never become valid again. If we
+                // wanted to make the rule hold more strongly, then sav_rhs_handle could return a
+                // fully-null handle when the sav_rhs column is disabled, but extra work would be
+                // needed to ensure that (enable column, get handle, disable column, enable column)
+                // consistently left an *invalid* handle.
+                CHECK(handle);
+            }
+        }
+    }
+    GIVEN("fast_imem calculation is enabled") {
+        set_fast_imem(true);
+        WHEN("A node is default-constructed") {
+            REQUIRE(neuron::model().node_data().size() == 0);
+            ::Node node{};
+            check_default(node);
+            auto handle = node.sav_rhs_handle();
+            *handle = 42;  // non-default value
+            generic_data_handle generic{handle};
+            CHECK(handle);
+            CHECK(to_str(handle) ==
+                  "data_handle<double>{Node::field::FastIMemSavRHS row=0/1 val=42}");
+            CHECK(to_str(generic) ==
+                  "generic_data_handle{Node::field::FastIMemSavRHS row=0/1 type=double*}");
+            AND_WHEN("fast_imem calculation is disabled with a Node active") {
+                REQUIRE(neuron::model().node_data().size() == 1);
+                set_fast_imem(false);
+                check_throws(node);
+                // This handle used to be valid, but it is now invalid because the optional field it
+                // refers to has been disabled.
+                CHECK_FALSE(handle);
+                CHECK(to_str(handle) ==
+                      "data_handle<double>{Node::field::FastIMemSavRHS row=0/0 disabled}");
+                CHECK(to_str(generic) ==
+                      "generic_data_handle{Node::field::FastIMemSavRHS row=0/0 disabled "
+                      "type=double*}");
+                AND_WHEN("fast_imem calculation is re-enabled") {
+                    set_fast_imem(true);
+                    // non-default value written above has been lost
+                    check_default(node);
+                    // The relevant optional column has been enabled again, so the handles are valid
+                    // once more.
+                    CHECK(handle);
+                    CHECK(to_str(handle) ==
+                          "data_handle<double>{Node::field::FastIMemSavRHS row=0/1 val=0}");
```

Comment:
There were some choices to be made here regarding the interaction of optional columns (`sav_d`, `sav_rhs` in the Node case) and the data handles.

The current approach is that the handle you get by calling
```c++
node.sav_rhs_handle()
```
when the `sav_rhs` column is **disabled** is invalid, in the sense that it evaluates to `false` and throws if you try and dereference it, but nonetheless **does** known which Node it refers to, and that it refers to the `sav_rhs` column.

This means that if you enable the `sav_rhs` column (by calling `CVode.use_fast_imem`) then the handle magically/seamlessly becomes valid. In other words, the bookkeeping-related stuff all exists for disabled-at-runtime columns, the only thing that is enabled/disabled is the actual data storage.

This behaviour was fairly natural/easy to achieve, but it does violate the otherwise-true (I believe) rule that a data handle that is invalid (evaluates to false) will never spontaneously become valid (evaluate to true, be dereferenced).

The alternative where `sav_rhs_handle()` only returns a valid handle if the `sav_rhs` column is enabled, and where that handle only remains valid so long as the `sav_rhs` column **stays** enabled, is also viable. In that case, the pretty-printing would lose some of its abilities.

Reply:
cc: @nrnhines @iomaganaris for any thoughts

Reply:
On reflection and on balance, I think the alternative/simpler version is easier to reason about/explain, so I switched to that.